activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb

```diff
@@ -89,7 +89,7 @@ def initialize(...)
             end
         end
 
-        def checkout(...)
+        def lease_connection
```

Comment:
We now only set the query cache for leased connection. If you do a raw checkout, there's no query cache, which I think makes more sense because assigning the same cache to two connections could lead to inconsistency if one connection is in a transaction and the other not.

Reply:
Also I think it would make sense to move the `QueryCache::Store` inside the `Lease`, this way we have a single object that hold all the thread specific state.

But I'm trying to keep this PR at a reasonable size.